kafka/client_async.py

```diff
@@ -198,6 +199,7 @@ def __init__(self, **configs):
         self._bootstrap_fails = 0
         self._wake_r, self._wake_w = socket.socketpair()
         self._wake_r.setblocking(False)
+        self._wake_w.settimeout(self.config['max_block_ms'] / 1000.0)
```

Comment:
I remember we set setblocking(True) in code sometimes. That operation resets timeout if I remember correctly. Could you make sure we don't do it on this socket?

Reply:
yes, here is all the `setblocking()` calls in 1.4.3:

```
$ git grep 'setblocking'
kafka/client_async.py:        self._wake_r.setblocking(False)
kafka/conn.py:            self._sock.setblocking(False)
kafka/vendor/socketpair.py:        # setblocking(False) that prevents us from having to create a thread.
kafka/vendor/socketpair.py:                csock.setblocking(False)
kafka/vendor/socketpair.py:                csock.setblocking(True)
```

the csock in socketpair.py is setblocking(True) before returning as `self._wake_w` in `KafkaClient.__init__()`. 

```
from kafka import KafkaProducer
import logging
import sys

logging.basicConfig(stream=sys.stdout, level=logging.INFO)

KAFKA_CONFIG = {
    'batch_size': 163840,
    'linger_ms': 10000,  # enlarge the linger_ms to run into the batch_is_full condition
    'max_block_ms': 10,
    'reconnect_backoff_ms': 300,
}

producer = KafkaProducer(bootstrap_servers='kafka03.dev.rack.xxxxxxx.com:3390', **KAFKA_CONFIG)
i = 0

while True:
    producer.send('log.franz.test', b'some_message_bytes')
    i += 1
    if i % 1000 == 0:
        print('1000')

```

this is the script to reproduce the hangs, when this script is looping, close the wifi connection. and we will see it hangs, and get this traceback on pressing ^C:

<img width="589" alt="screen shot 2018-08-21 at 10 52 05 am" src="https://user-images.githubusercontent.com/129800/44377657-59e51a00-a530-11e8-9905-57183f97efe7.png">

after setting this socket timeout, this script can quit immediately after `max_block_ms`.